### PR TITLE
Add Admission webhook

### PR DESCRIPTION
**Pull request by jingxu97**
This PR adds the admission webhook for verifying snapshot API objects
and also setting the default snapshot class if it is not specified by the
users

### CODE REVIEW

File: cmd/csi-snapshotter/main.go [hunk 1]
```diff
@@ -22,9 +22,13 @@ import (
 	"fmt"
 	"os"
 	"os/signal"
+	//"strconv"
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go]: Can you remove this since it is not needed?

File: cmd/csi-snapshotter/main.go [hunk 2]
```diff
@@ -22,9 +22,13 @@ import (
...
 	"time"
+	//"net"
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go]: Same here

File: cmd/csi-snapshotter/main.go [hunk 3]
```diff
@@ -33,9 +37,11 @@ import (
 	"github.com/kubernetes-csi/external-snapshotter/pkg/connection"
 	"github.com/kubernetes-csi/external-snapshotter/pkg/controller"
 
+	//k8snet "k8s.io/apimachinery/pkg/util/net"
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go]: Same here

File: cmd/csi-snapshotter/main.go [hunk 4]
```diff
@@ -203,3 +215,23 @@ func waitForDriverReady(csiConn connection.CSIConnection, timeout time.Duration)
 		time.Sleep(time.Second)
 	}
 }
+
+func startWebhooksHTTPs(kubeClientset kubernetes.Interface, snapClient clientset.Interface) error {
+	glog.Infof("Starting Snapshot webhooks HTTPS handler")
+	defer glog.Infof("Stopping Snapshot webhooks HTTPS handler")
```
Comment (xing-yang) [inline cmd/csi-snapshotter/main.go]: Please use V(5) for debug log msg.

File: hack/gencerts.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
+
+#!/bin/bash
+
+# Copyright 2015 Google Inc. All rights reserved.
```
Comment (xing-yang) [inline hack/gencerts.sh]: 2018?

Also should we use "Copyright 2018 The Kubernetes Authors." to be consistent with copyrights in other files?

File: hack/gencerts.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# Create a server certificate
+openssl genrsa -out serverKey.pem 2048
+# Note the CN is the DNS name of the service of the webhook.
+openssl req -new -key serverKey.pem -out server.csr -subj "/CN=${CN}" -config server.conf
+openssl x509 -req -in server.csr -CA caCert.pem -CAkey caKey.pem -CAcreateserial -out serverCert.pem -days 100000 -extensions v3_req -extfile server.conf
+
+outfile=pkg/webhook/certs.go
+
+cat > $outfile << EOF
+// Copyright 2018 Google Inc. All Rights Reserved.
```
Comment (xing-yang) [inline hack/gencerts.sh]: Is copyright needed here?

File: pkg/webhook/certs.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
```
Comment (xing-yang) [inline pkg/webhook/certs.go]: Same question about copyright here and other places.

File: pkg/webhook/webhook.go [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mattbaird/jsonpatch"
+	//"github.com/pkg/errors"
+	//"github.com/sirupsen/logrus"
```
Comment (xing-yang) [inline pkg/webhook/webhook.go]: Remove lines that are commented out.

File: pkg/webhook/webhook.go [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+	clientset "github.com/kubernetes-csi/external-snapshotter/pkg/client/clientset/versioned"
+)
+
+const (
+	webhookConfigName   = "snapshot-webhook-config"
+	snapshotWebhookName = "snapshot-init.snapshot.storage.k8s.com"
+	// The admission webhook must be exposed in the following service,
+	// this is mainly for the server certificate.
+	// serviceName is the name of the admission webhook service.
+	//serviceName = "snapshot-webhook-svc"
```
Comment (xing-yang) [inline pkg/webhook/webhook.go]: Remove this line

File: pkg/webhook/webhook.go [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+							APIVersions: []string{"v1alpha1"},
+							Resources:   []string{"volumesnapshots"},
+						},
+					},
+				},
+				FailurePolicy: &failurePolicy,
+			},
+		},
+	}
+	glog.V(2).Infof("Creating ValidatingWebhookConfiguration %q", config.Name)
```
Comment (xing-yang) [inline pkg/webhook/webhook.go]: Should this be a V(5) for debug msg?

File: pkg/webhook/webhook.go [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+	if err != nil {
+		review.Response.Result = &metav1.Status{
+			Reason:  metav1.StatusReasonInvalid,
+			Message: err.Error(),
+		}
+		return review
+	}
+
+	if changed {
+		glog.V(4).Infof("Patched Snapshot %s/%s: +%v", snapshot.Namespace, snapshot.Name, newSnapshot)
```
Comment (xing-yang) [inline pkg/webhook/webhook.go]: V(5) for debug msg?

File: pkg/webhook/webhook.go [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+			return nil, false, err
+		}
+		snapshotClone := snapshot.DeepCopy()
+		snapshotClone.Spec.VolumeSnapshotClassName = &defaultClassName
+		return snapshotClone, true, nil
+	}
+
+	// if source is not specified, the content name must be set
+	if snapshot.Spec.SnapshotContentName == "" {
+		return nil, false, fmt.Errorf("Cannot set Snapshot.Spec.Source to nil and snapshot.Spec.SnapshotContentName to empty at the same time.")
```
Comment (xing-yang) [inline pkg/webhook/webhook.go]: lower case to start an error: s/Cannot/cannot

File: pkg/webhook/webhook.go [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+		return true, storageClass.Provisioner, fmt.Errorf("the snapshotter driver specified in snapshot class does not match the driver from the volume's provisioner")
+	}
+	return true, storageClass.Provisioner, nil
+}
+
+// sendResponse sends the response using the writer.
+func sendResponse(writer http.ResponseWriter, response interface{}) {
+	b, err := json.Marshal(response)
+	if err != nil {
+		//logrus.WithError(err).Errorf("Failed to marshal response: %+v", response)
```
Comment (xing-yang) [inline pkg/webhook/webhook.go]: Remove this line.

File: cmd/csi-snapshotter/main.go (hunk 1 above)
Comment (jingxu97) [reply cmd/csi-snapshotter/main.go]: fixed.

File: cmd/csi-snapshotter/main.go (hunk 2 above)
Comment (jingxu97) [reply cmd/csi-snapshotter/main.go]: fixed

File: cmd/csi-snapshotter/main.go (hunk 3 above)
Comment (jingxu97) [reply cmd/csi-snapshotter/main.go]: fixed

File: cmd/csi-snapshotter/main.go (hunk 4 above)
Comment (jingxu97) [reply cmd/csi-snapshotter/main.go]: I added V(2) since it only shows up once.

File: hack/gencerts.sh (hunk 5 above)
Comment (jingxu97) [reply hack/gencerts.sh]: fixed

File: pkg/webhook/webhook.go (hunk 9 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go (hunk 10 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: I think it is ok since it only shows up once.

File: pkg/webhook/webhook.go (hunk 11 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: changed to V(5)

File: pkg/webhook/webhook.go (hunk 13 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go (hunk 12 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+	serviceName = "snapshot-admission-webhook-svc"
+
+	IsDefaultSnapshotClassAnnotation = "snapshot.storage.kubernetes.io/is-default-class"
+)
+
+var supportedSourceKinds = sets.NewString(string("PersistentVolumeClaim"))
+var supportedDataSourceAPIGroups = sets.NewString(string(""))
+
+// CreateConfiguration creates MutatingWebhookConfiguration and registeres the
+// webhood admission controller with the kube-apiserver
```
[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+							APIVersions: []string{"v1alpha1"},
+							Resources:   []string{"volumesnapshots"},
+						},
+					},
+				},
+				FailurePolicy: &failurePolicy,
+			},
+		},
+	}
+	glog.V(2).Infof("Creating ValidatingWebhookConfiguration %q", config.Name)
```
[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+	if _, err := clientset.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Create(config); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			glog.V(2).Infof("ValidatingWebhookConfiguration %q already exists; use the existing one", config.Name)
```
[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+			return nil
+		}
+		return fmt.Errorf("failed to create ValidatingWebhookConfiguration %q: %v", config.Name, err)
```
[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+	"k8s.io/apimachinery/pkg/util/sets"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	snapshotv1alpha1 "github.com/kubernetes-csi/external-snapshotter/pkg/apis/volumesnapshot/v1alpha1"
+	clientset "github.com/kubernetes-csi/external-snapshotter/pkg/client/clientset/versioned"
+)
+
+const (
+	webhookConfigName   = "snapshot-webhook-config"
+	snapshotWebhookName = "snapshot-init.snapshot.storage.k8s.com"
```
[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/webhook/webhook.go]: changed to .io. Thanks!

**Comment by jingxu97** [comment]
Addressed comments. PTAL. Thanks!

**Comment by xing-yang** [comment]
Need a rebase?

File: pkg/webhook/webhook.go [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+
+	IsDefaultSnapshotClassAnnotation = "snapshot.storage.kubernetes.io/is-default-class"
+)
+
+var supportedSourceKinds = sets.NewString(string("PersistentVolumeClaim"))
+var supportedDataSourceAPIGroups = sets.NewString(string(""))
+
+// CreateConfiguration creates MutatingWebhookConfiguration and registeres the
+// webhook admission controller with the kube-apiserver
+func CreateConfiguration(clientset kubeclientset.Interface, url string) error {
```
Comment (wackxu) [inline pkg/webhook/webhook.go]: the `url` parameter is not used, we can  remove it. or use it and update the above `mux.Handle`

File: pkg/webhook/webhook.go [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+		if apierrors.IsAlreadyExists(err) {
+			glog.V(2).Infof("MutatingWebhookConfigurations %q already exists; use the existing one", config.Name)
+			return nil
+		}
+		return fmt.Errorf("failed to create MutatingWebhookConfigurations %q: %v", config.Name, err)
+	}
+	return nil
+}
+
+func GetTLSConfig(clientset kubeclientset.Interface) *tls.Config {
```
Comment (wackxu) [inline pkg/webhook/webhook.go]: `clientset ` parameter not used

File: pkg/webhook/webhook.go (hunk 19 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

File: pkg/webhook/webhook.go (hunk 20 above)
Comment (jingxu97) [reply pkg/webhook/webhook.go]: fixed

**Comment by xing-yang** [comment]
Can you please rebase this?  Thanks.

File: hack/gencerts.sh (hunk 6 above)
Comment (xing-yang) [reply hack/gencerts.sh]: This copyright section seems out of place.  Is it needed?  Also should it be "Kubernetes Authors" if we need copyright?

**Comment by jingxu97** [comment]
rebased. PTAL

**Comment by xing-yang** [comment]
CI failed.

**Comment by wackxu** [comment]
The CI is green now,  I think we have several things to do:
1. We should not use the sample certificate in the ` pkg/webhook/certs.go` in product environment, we should add some flags to pass users key and  certificate.
2. the register MutatingWebhookConfiguration is point to a ServiceReference. we should add the yaml to deploy the webhook controller, such as service, deployment.
3. We should clean up the repeated logic in the controller that already exist in the admission controller.

@xing-yang @jingxu97 Any thoughts?

**Comment by wackxu** [comment]
@jingxu97 Can you update the code, we should push this forward.

**Pull request closed without merging**